pkg/apis/pipeline/v1alpha1/run_types.go

```diff
@@ -41,6 +41,10 @@ type RunSpec struct {
 	// +optional
 	Ref *TaskRef `json:"ref,omitempty"`
 
+	// Spec is a specification of a custom task
+	// +optional
+	Spec *v1beta1.EmbeddedTask `json:"spec,omitempty"`
```

Comment:
Ah, something I had not realised during TEP review is that this also means we're nesting `TaskSpec` fields (like pipeline resources etc,) and `PipelineTaskMetadata` fields (annotations, labels) into the `RunSpec`.

Is there a reason to re-use `EmbeddedTask` here?  Or could we introduce a new type that only includes the fields we want for `RunSpec`?  E.g.

```go
type EmbeddedRunSpec struct {
	// +optional
	runtime.TypeMeta `json:",inline,omitempty"`

	// Spec is a specification of a custom task
	// +optional
	Spec runtime.RawExtension `json:"spec,omitempty"`
}
```

Reply:
Sorry, another question after reviewing this with the pipelines owners group - what is the purpose of `runtime.TypeMeta` in this location?  Would this potentially differ from the `run.apiVersion` and `run.kind`?

Reply:
>Sorry, another question after reviewing this with the pipelines owners group - what is the purpose of runtime.TypeMeta in this location? Would this potentially differ from the run.apiVersion and run.kind? 

It will not differ from `run.apiVersion` and `run.kind`, my guess is for brevity we prefer this.

>Is there a reason to re-use EmbeddedTask here? Or could we introduce a new type that only includes the fields we want for RunSpec? E.g.

Intention is, custom tasks will become consistent with regular tasks at some point. e.g. we need `PipelineTaskMetadata` to provide label to embedded task. 
If you still feel otherwise, let us know.

Reply:
## TypeMeta

Ah, ok.  Going back to the docs in this example helped clear things up for me wrt TypeMeta:

```
apiVersion: tekton.dev/v1alpha1
kind: Run
metadata:
  name: embedded-run
spec:
  spec:
    apiVersion: example.dev/v1alpha1
    kind: Example
    spec:
      field1: value1
      field2: value2
```

The run's API version and the spec's API version can differ - one describes the tekton resource "tekton.dev/v1alpha1.Run" and the other describes the Custom Task's own apiVersion and kind.  I find this distinction _endlessly_ confusing but that's just the way it is with Custom Tasks. Apologies for muddying things up here.  TypeMeta is fine as-is.

---

## EmbeddedTask

I do want to avoid using `EmbeddedTask` if we can for the following reasons:
- It brings baggage with it that we don't have a concrete use case for yet in Runs (e.g. pipeline resources).
- All of the fields from `TaskSpec` are already representable by virtue of using `runtime.RawExtension` I think?
- We can make `EmbeddedRunSpec` a subset of fields from `EmbeddedTask` and eventually converge them if that's a goal we have.  Although I would also say here that I wasn't aware of this being a Goal - is it mentioned in the TEP or a comment on it?  I probably missed it.

So to support metadata fields as well it sounds like the shape would become:

```go
type RunSpecMetadata struct {
	// +optional
	Labels map[string]string `json:"labels,omitempty"`

	// +optional
	Annotations map[string]string `json:"annotations,omitempty"`
}

type EmbeddedRunSpec struct {
	// +optional
	runtime.TypeMeta `json:",inline,omitempty"`

	// +optional
	Metadata RunSpecMetadata `json:"metadata,omitempty"`

	// Spec is a specification of a custom task
	// +optional
	Spec runtime.RawExtension `json:"spec,omitempty"`
}
```

The `createRun` func becomes a little more complicated because it has to copy values from the `EmbeddedTask` to the `EmbeddedRunSpec` but I much prefer we start with a constrained set of fields and expand it when needed.  WDYT?

Reply:
Alright, agree with you, do you think we would need to update the TEP with above suggestion for types?

Reply:
And get it approved?

Reply:
@sbwsg At the moment a `PipelineTask` has a nested `EmbeddedTask` and as a result, the custom task can be expressed under the field `taskSpec`. After this change, how will that change? 

Few options:
1. We introduce a new optional field `spec` to store the `EmbeddedRunSpec` inside a `PipelineTask`. Downside, two different field with similar sounding field names and purposes.
2. Leave current `EmbeddedTask` as is, and introduce a new type `EmbeddedRunSpec` for `Run`. Downside, same fields copied twice.
3. Any other suggestions?

Reply:
Once this PR is merged the TEP can be moved to `"implemented"` state.  My suggestion would be to include these changes as part of that PR.

Also pinging @tektoncd/core-maintainers for any other opinions / feedback on the desired process here.

Reply:
> Leave current EmbeddedTask as is, and introduce a new type EmbeddedRunSpec for Run. Downside, same fields copied twice.

This would definitely be my preference.

Reply:
Yes, indeed.

Reply:
@sbwsg One question, now that `Metadata` field is shifted from `EmbeddedTask` to `EmbeddedSpec` which is inlined into `EmbeddedTask`, is it a matter of concern?